extensions/VisualEditor/wikia/modules/ve/dm/ve.dm.WikiaTransclusionModel.js

```diff
@@ -2,148 +2,188 @@
+		};
 
-	if ( data && Object.keys( data.query ).length && Object.keys( data.query.pages ).length ) {
-		denormalizedData = this.denormalizeInfoboxTemplateTitles( data );
+		if ( sourceMetadata.primary === true ) {
+			specs.params[sourceName].type = node.type;
+		}
 
-		for ( id in denormalizedData ) {
-			page = denormalizedData[id];
+		specs.paramOrder.push( sourceName );
+	}
 
-			if ( !page.infoboxes ) {
-				return;
-			}
+	/**
+	 * @param {Object} node
+	 * @param {Object} specs
+	 */
+	function parseInfoboxMetadata( node, specs ) {
+		if ( node.type === 'group' ) {
+			node.metadata.forEach( function ( childNode ) {
+				parseInfoboxMetadata( childNode, specs );
+			} );
+		} else {
+			Object.keys( node.sources ).map( function ( sourceName ) {
+				var sourceMetadata = node.sources[sourceName];
+				addInfoboxSourceParam( specs, node, sourceName, sourceMetadata );
+			} );
+		}
+	}
 
-			if ( !specs[page.title] ) {
-				specs[page.title] = {
-					title: page.title,
-					description: '',
-					params: {},
-					paramOrder: []
-				};
-			}
+	/**
+	 * @desc process the infobox params received from API. For all requested pages, if they contain infoboxes,
+	 * extend their param array with received infobox params.
+	 */
+	ve.dm.WikiaTransclusionModel.prototype.fetchInfoboxParamsRequestDone = function ( specs, data ) {
+		var id, page, denormalizedData;
 
-			for ( i = 0; i < page.infoboxes.length; i++ ) {
-				page.infoboxes[i].sources = Object.keys( page.infoboxes[i].sourcelabels );
+		if ( data && Object.keys( data.query ).length && Object.keys( data.query.pages ).length ) {
+			denormalizedData = this.denormalizeInfoboxTemplateTitles( data );
 
-				for ( j = 0; j < page.infoboxes[i].sources.length; j++ ) {
-					source = page.infoboxes[i].sources[j];
+			for ( id in denormalizedData ) {
+				page = denormalizedData[id];
 
-					specs[page.title].params[ source ] = { label: page.infoboxes[i].sourcelabels[source] };
-					specs[page.title].paramOrder.push( page.infoboxes[i].sources[j] );
+				if ( !page.infoboxes ) {
+					return;
 				}
-			}
-		}
 
-		ve.extendObject( this.specCache, specs );
-	}
-};
-
-/**
- * @desc The template names which contain _ are normalized in infobox API and the title from response contain spaces.
- * It's inconsistent with the template API. However, the response from infobox API contains field 'normalized'
- * where we can take the requested title from.
- *
- * @param data from infobox params API
- */
-ve.dm.WikiaTransclusionModel.prototype.denormalizeInfoboxTemplateTitles = function ( data ) {
-	var i, id, title, pages = data.query.pages;
+				if ( !specs[page.title] ) {
+					specs[page.title] = {
+						title: page.title,
+						description: '',
+						params: {},
+						paramOrder: []
+					};
+				}
 
-	if ( data.query.normalized ) {
-		for ( i = 0; i < data.query.normalized.length; i++ ) {
-			title = data.query.normalized[i];
+				page.infoboxes.forEach( function ( infobox ) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Define this function outside of a loop. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AFunctionDefinitionInsideLoop)